File: GFG/love_babbar_dsa_cheatsheet/backtracking/knights_tour.cpp
#include <bits/stdc++.h>
using namespace std;

// Define knight's possible moves
const vector<pair<int, int>> knightMoves = {
    {2, 1}, {2, -1}, {-2, 1}, {-2, -1},
    {1, 2}, {1, -2}, {-1, 2}, {-1, -2}
};
int answer[8][8]; // Global array to store the final solution

// Function to copy ChessBoard to answer
void copyBoard(int ChessBoard[8][8], int answer[8][8]) {
    for (int i = 0; i < 8; i++) {
        for (int j = 0; j < 8; j++) {
            answer[i][j] = ChessBoard[i][j];
        }
    }
}

// Function to perform DFS and solve the Knight's Tour
bool dfs(int x, int y, int ChessBoard[8][8], set<pair<int, int>> &visited) {
    if (visited.size() == 64) {
        // Base case: all cells are visited
        /*copyBoard(ChessBoard, answer); // Copy solution to the global answer*/
        return true;
    }

    // Check bounds and whether cell is already visited
    if (x < 0 || y < 0 || x >= 8 || y >= 8 || visited.count({x, y})) {
        return false;
    }

    // Mark the current cell as visited and update ChessBoard
    visited.insert({x, y});
    ChessBoard[x][y] = visited.size();

    // Explore all possible knight moves
    for (auto move : knightMoves) {
        int nextX = x + move.first;
        int nextY = y + move.second;
        if (dfs(nextX, nextY, ChessBoard, visited)) {
            return true; // Solution found
        }
    }

    // Backtracking: unmark the current cell and reset ChessBoard
    visited.erase({x, y});
    ChessBoard[x][y] = 0;

    return false; // No valid moves from this point
}

int main() {
    // Initialize the ChessBoard
    int ChessBoard[8][8];
    for (int i = 0; i < 8; i++) {
        for (int j = 0; j < 8; j++) {
            ChessBoard[i][j] = 0;
        }
    }

    // Set to track visited cells
    set<pair<int, int>> visited;

    // Start DFS from the top-left corner
    if (dfs(0, 0, ChessBoard, visited)) {
        // Print the solution from the global answer array
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                cout << setw(3) << answer[i][j] << " ";
            }
            cout << endl;
        }
    } else {
        cout << "No solution exists for the Knight's Tour!" << endl;
    }

    return 0;
}

File: GFG/love_babbar_dsa_cheatsheet/backtracking/knights_tour.py
import time


def initialize_board(size):
    """Initialize an empty chessboard with -1 to indicate unvisited squares."""
    return [[-1 for _ in range(size)] for _ in range(size)]


def is_valid_move(x, y, board):
    """Check if the move is valid: within bounds and not yet visited."""
    size = len(board)
    return 0 <= x < size and 0 <= y < size and board[x][y] == -1


def solve_knights_tour(board, x, y, move_count, start_time, time_limit):
    """Recursively attempt to solve the Knight's Tour problem using backtracking."""
    size = len(board)

    # Check if the time limit is exceeded
    if time.time() - start_time > time_limit:
        return False

    if move_count == size * size:
        return True

    knight_moves = [
        (2, 1),
        (1, 2),
        (-1, 2),
        (-2, 1),
        (-2, -1),
        (-1, -2),
        (1, -2),
        (2, -1),
    ]

    for dx, dy in knight_moves:
        new_x, new_y = x + dx, y + dy
        if is_valid_move(new_x, new_y, board):
            board[new_x][new_y] = move_count
            if solve_knights_tour(
                board, new_x, new_y, move_count + 1, start_time, time_limit
            ):
                return True
            board[new_x][new_y] = -1  # Backtrack
    return False


def print_board(board):
    """Print the chessboard with the knight's path."""
    for row in board:
        print(" ".join(str(cell).rjust(3, " ") for cell in row))


def main():
    size = 8  # Size of the chessboard
    board = initialize_board(size)

    start_x, start_y = 0, 0  # Starting position of the knight
    board[start_x][start_y] = 0  # Initialize the starting position with the first move

    time_limit = 10  # Maximum time (in seconds) to search for a solution
    start_time = time.time()

    if solve_knights_tour(board, start_x, start_y, 1, start_time, time_limit):
        print("Solution found!")
        print_board(board)
    else:
        print("No solution exists or the search exceeded the time limit.")


if __name__ == "__main__":
    main()
